promise/src/main/java/com/iluwatar/promise/Promise.java

```diff
@@ -0,0 +1,170 @@
+import java.util.concurrent.Executor;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * Implements the promise pattern.
+ * 
+ * @param <T> type of result.
+ */
+public class Promise<T> extends PromiseSupport<T> {
+
+  private Runnable fulfillmentAction;
+
+  /**
+   * Creates a promise that will be fulfilled in future.
+   */
+  public Promise() {
+  }
+
+  /**
+   * Fulfills the promise with the provided value.
+   * @param value the fulfilled value that can be accessed using {@link #get()}.
+   */
+  @Override
+  public void fulfill(T value) {
+    super.fulfill(value);
+    postFulfillment();
+  }
+
+  /**
+   * Fulfills the promise with exception due to error in execution.
+   * @param exception the exception will be wrapped in {@link ExecutionException}
+   *        when accessing the value using {@link #get()}.
+   */
+  @Override
+  public void fulfillExceptionally(Exception exception) {
+    super.fulfillExceptionally(exception);
+    postFulfillment();
+  }
+
+  private void postFulfillment() {
+    if (fulfillmentAction == null) {
+      return;
+    }
+    fulfillmentAction.run();
+  }
+
+  /**
+   * Executes the task using the executor in other thread and fulfills the promise returned
+   * once the task completes either successfully or with an exception.
+   * 
+   * @param task the task that will provide the value to fulfill the promise.
+   * @param executor the executor in which the task should be run.
+   * @return a promise that represents the result of running the task provided.
+   */
+  public Promise<T> fulfillInAsync(final Callable<T> task, Executor executor) {
+    executor.execute(() -> {
+      try {
+        fulfill(task.call());
+      } catch (Exception e) {
+        fulfillExceptionally(e);
+      }
+    });
+    return this;
+  }
+
+  /**
+   * Returns a new promise that, when this promise is fulfilled normally, is fulfilled with 
+   * result of this promise as argument to the action provided.
+   * @param action action to be executed.
+   * @return a new promise.
+   */
+  public Promise<Void> then(Consumer<? super T> action) {
+    Promise<Void> dest = new Promise<>();
+    fulfillmentAction = new ConsumeAction(this, dest, action);
+    return dest;
+  }
+
+  /**
+   * Returns a new promise that, when this promise is fulfilled normally, is fulfilled with 
+   * result of this promise as argument to the function provided.
+   * @param func function to be executed.
+   * @return a new promise.
+   */
+  public <V> Promise<V> then(Function<? super T, V> func) {
+    Promise<V> dest = new Promise<>();
+    fulfillmentAction = new TransformAction<V>(this, dest, func);
+    return dest;
+  }
+
+  /**
+   * A consume action provides the action, the value from source promise and fulfills the
+   * destination promise.
+   */
+  private class ConsumeAction implements Runnable {
+
+    private Promise<T> src;
+    private Promise<Void> dest;
+    private Consumer<? super T> action;
```

Comment:
The private fields could be declared final since they are only set in the constructor.